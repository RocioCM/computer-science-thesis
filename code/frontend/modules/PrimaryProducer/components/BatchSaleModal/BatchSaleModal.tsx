import Modal, { ModalProps } from '@/common/components/Modal';

interface Props extends ModalProps {
  handleCancel: () => any;
  editingId: number | null;
  handleSuccess: () => void;
}

const BatchSaleModal: React.FC<Props> = ({
  editingId,
  handleCancel,
  handleSuccess,
  ...props
}) => {
  return (
    <Modal handleCancel={handleCancel} big skippable={false} {...props}>
      <h2 className="w-full bg-n0">Vender lote de botellas #{editingId}</h2>
    </Modal>
  );
};

export default BatchSaleModal;
