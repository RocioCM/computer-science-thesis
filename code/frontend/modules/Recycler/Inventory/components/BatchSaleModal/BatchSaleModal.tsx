import Modal, { ModalProps } from '@/common/components/Modal';
import RecyclerServices from '../../services';
import { SellRecyclingBatchPayload, Buyer } from '../../types';
import { toast } from 'react-toastify';
import Button from '@/common/components/Button';
import { SELL_FORM_INPUTS, SELL_FORM_STRUCT } from '../../constants';
import useForm from '@/common/hooks/useForm';

interface Props extends ModalProps {
  handleCancel: () => any;
  editingId: number | null;
  handleSuccess: () => void;
}

const BatchSaleModal: React.FC<Props> = ({
  editingId,
  handleCancel,
  handleSuccess,
  ...props
}) => {
  const { form, submitEnabled, formBuilder } = useForm(SELL_FORM_STRUCT);

  const getBuyers = async (query: string) => {
    const { data } = await RecyclerServices.searchBuyers(query);
    return (
      data?.map((buyer: Buyer) => ({
        label: `${buyer.userName}${buyer.userName ? ' - ' : ''}${buyer.email}`,
        value: buyer.firebaseUid,
      })) ?? []
    );
  };

  const handleSale = async () => {
    const payload: SellRecyclingBatchPayload = {
      batchId: editingId as number,
      buyerUid: form.buyerUid,
    };
    const { ok } = await RecyclerServices.sellRecyclingBatch(payload);
    if (ok) {
      toast.success(
        'Venta completada exitosamente. El lote reciclado ya está en propiedad del comprador'
      );
      handleCancel();
      handleSuccess();
    } else {
      toast.error(
        'No se pudo completar la venta. Por favor, inténtelo de nuevo más tarde'
      );
    }
  };

  return (
    <Modal handleCancel={handleCancel} {...props}>
      <h2 className="w-full bg-n0">Venta lote #{editingId}</h2>
      <div className="w-full flex flex-col gap-m my-xl">
        {formBuilder(SELL_FORM_INPUTS, {
          handleSearch: { buyerUid: getBuyers },
        })}
      </div>
      <Button
        handleClick={handleSale}
        disabled={!submitEnabled}
        width="full"
        label="Vender"
      />
    </Modal>
  );
};

export default BatchSaleModal;
