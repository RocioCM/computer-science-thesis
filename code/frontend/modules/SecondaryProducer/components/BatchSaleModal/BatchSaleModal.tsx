import Modal, { ModalProps } from '@/common/components/Modal';
import SecondaryProducerServices from '../../services';
import { SellProductBottlesPayload, Buyer } from '../../types';
import { toast } from 'react-toastify';
import Button from '@/common/components/Button';
import { SELL_FORM_INPUTS, SELL_FORM_STRUCT } from '../../constants';
import useForm from '@/common/hooks/useForm';

interface Props extends ModalProps {
  handleCancel: () => any;
  editingId: number | null;
  handleSuccess: () => void;
}

const BatchSaleModal: React.FC<Props> = ({
  editingId,
  handleCancel,
  handleSuccess,
  ...props
}) => {
  const { form, submitEnabled, formBuilder } = useForm(SELL_FORM_STRUCT);

  const getBuyers = async (query: string) => {
    const { data } = await SecondaryProducerServices.searchBuyers(query);
    return (
      data?.map((buyer: Buyer) => ({
        label: `${buyer.userName}${buyer.userName ? ' - ' : ''}${buyer.email}`,
        value: buyer.firebaseUid,
      })) ?? []
    );
  };

  const handleSale = async () => {
    const payload: SellProductBottlesPayload = {
      batchId: editingId as number,
      quantity: form.quantity,
      buyerUid: form.buyerUid,
    };
    const { ok } = await SecondaryProducerServices.sellBatch(payload);
    if (ok) {
      toast.success(
        'Venta completada exitosamente. Los productos seleccionados ya están en propiedad del comprador'
      );
      handleCancel();
      handleSuccess();
    } else {
      toast.error(
        'No se pudo completar la venta. Por favor, inténtelo de nuevo más tarde'
      );
    }
  };

  return (
    <Modal handleCancel={handleCancel} {...props}>
      <h2 className="w-full bg-n0">Venta lote #{editingId}</h2>
      <div className="w-full flex flex-col gap-m my-xl">
        {formBuilder(SELL_FORM_INPUTS, {
          handleSearch: { buyerUid: getBuyers },
        })}
      </div>
      <Button
        handleClick={handleSale}
        disabled={!submitEnabled}
        width="full"
        label="Vender"
      />
    </Modal>
  );
};

export default BatchSaleModal;
