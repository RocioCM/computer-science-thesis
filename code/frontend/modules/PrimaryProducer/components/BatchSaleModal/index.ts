import BatchSaleModal from './BatchSaleModal';

export default BatchSaleModal;
