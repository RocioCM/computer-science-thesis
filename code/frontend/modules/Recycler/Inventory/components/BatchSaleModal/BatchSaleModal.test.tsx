import { render } from '@testing-library/react';
import BatchSaleModal from './BatchSaleModal';

describe('Recycler Inventory - BatchSaleModal', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Matches snapshot', () => {
    const { container } = render(
      <div id="modal-root">
        <BatchSaleModal
          editingId={1}
          handleCancel={() => {}}
          handleSuccess={() => {}}
        />
      </div>
    );
    expect(container).toMatchSnapshot();
  });
});
