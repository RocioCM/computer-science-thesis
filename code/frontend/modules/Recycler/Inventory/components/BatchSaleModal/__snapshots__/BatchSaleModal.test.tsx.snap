// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recycler Inventory - BatchSaleModal Matches snapshot 1`] = `
<div>
  <div
    id="modal-root"
  >
    <div
      class="bg-n10 bg-opacity-70 py-2xl px-xl w-full h-screen fixed top-0 left-0 z-modal flex justify-center items-center animate__animated animate__fadeIn cursor-pointer"
      data-testid="modal"
    >
      <div
        class="bg-n0 min-w-[300px] min-h-[200px] rounded-rm overflow-y-auto flex flex-col items-center relative cursor-default animate__animated animate__zoomIn w-[380px] h-max max-h-full max-w-full px-3xl py-3xl"
      >
        <div
          class="flex justify-end w-full h-0 sticky top-0 z-10"
        >
          <img
            alt="Cerrar"
            class="h-6 w-auto cursor-pointer object-contain"
            data-nimg="1"
            decoding="async"
            height="30"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
            style="color: transparent;"
            width="30"
          />
        </div>
        <h2
          class="w-full font-medium text-[1.75rem] bg-n0 border-b border-n2 pb-2"
        >
          Venta lote #
          1
        </h2>
        <div
          class="w-full flex flex-col gap-m my-xl"
        >
          <div
            class="w-full"
          >
            <span
              class="block relative z-10 w-max transition-all text-p mb-xs top-0 left-0 text-n4 font-medium"
            >
              Comprador
            </span>
            <div
              class="relative flex flex-row flex-nowrap justify-between items-center w-full"
            >
              <div
                class="absolute left-0 h-full w-max flex flex-row items-center gap-xs"
              />
              <input
                class="input block w-full relative border bg-transparent outline-none transition-shadow p rounded-rs border-n3 px-m py-xs placeholder:text-n2 h-10 hover:border-n2 inputFocus focus:border-p1 focus:border-2 disabled:text-n3 disabled:bg-n1 disabled:border-n2 disabled:placeholder:text-transparent pr-8"
                data-testid="input"
                form="[object Object]"
                inputlabel="Comprador"
                name="buyerUid"
                options=""
                placeholder="Busca por nombre o correo del comprador"
                required=""
                type="text"
                value=""
              />
              <div
                class="absolute right-0 h-full w-max flex flex-row items-center gap-xs"
              >
                <i
                  class="fa-solid fa-search mr-3"
                  data-testid="fa-icon"
                />
              </div>
            </div>
          </div>
        </div>
        <button
          class="relative flex items-center justify-center gap-s rounded-rs font-semibold h-10 disabled:cursor-default disabled:bg-opacity-40  focusButton bg-p1 text-n0 py-s px-m w-full"
          data-testid="button"
          disabled=""
        >
          <span>
            Vender
          </span>
        </button>
      </div>
      <style>
        
          body {
            overflow: hidden;
          }
        
      </style>
    </div>
  </div>
</div>
`;
